### ISSUE TITLE
MMPose inference using MM Deploy

### ISSUE THREAD

**Issue by RS00001** [open-mmlab/mmdeploy#135]
I've been testing the MMDet and MMPose on Jetson Nano and the inference is very slow and so I've been meaning to test the inference on MM Deploy since it supports Jetson's TensorRT.

Could you please confirm if MM Pose is supported or not on MM Deploy?

**Comment by RunningLeon** [comment]
@RS00001 Hi, MMPose is planned to be supported in MMDeploy. Related PR is  https://github.com/open-mmlab/mmdeploy/pull/94.

**Comment by RS00001** [comment]
Understood. Will that work with the animal pose models?

**Comment by RunningLeon** [comment]
Currently only support mmpose deployment with following models:

- [configs/body/2d_kpt_sview_rgb_img/topdown_heatmap/coco/hrnet_w48_coco_256x192.py](https://github.com/open-mmlab/mmpose/blob/master/configs/body/2d_kpt_sview_rgb_img/topdown_heatmap/coco/hrnet_w48_coco_256x192.py)
- [configs/body/2d_kpt_sview_rgb_img/topdown_heatmap/coco/litehrnet_30_coco_256x192.py](https://github.com/open-mmlab/mmpose/blob/master/configs/body/2d_kpt_sview_rgb_img/topdown_heatmap/coco/litehrnet_30_coco_256x192.py)
- [configs/body/2d_kpt_sview_rgb_img/topdown_heatmap/coco/4xmspn50_coco_256x192.py](https://github.com/open-mmlab/mmpose/blob/master/configs/body/2d_kpt_sview_rgb_img/topdown_heatmap/coco/4xmspn50_coco_256x192.py)

**Comment by RunningLeon** [comment]
> Understood. Will that work with the animal pose models?

@RS00001 You could try the animal pose model once the PR is merged.

**Comment by RunningLeon** [comment]
> > I've been testing the MMDet and MMPose on Jetson Nano and the inference is very slow and so I've been meaning to test the inference on MM Deploy since it supports Jetson's TensorRT.
> > Could you please confirm if MM Pose is supported or not on MM Deploy?
> 
> Hi, I am trying to setup mmPose and mmdet in Jetson nano. I tried it in laptop. It worked fine. I think setting up both frameworks is totally different in Jetson nano as it uses aarch64. **Can you elaborate all steps for setting up mmdet and mmPose on Jetson nano. The help is much appreciated.**

You could refer to the doc: https://github.com/open-mmlab/mmdeploy/blob/master/docs/en/tutorials/how_to_install_mmdeploy_on_jetsons.md

**Comment by Thevakumar-Luheerathan** [comment]
> > > I've been testing the MMDet and MMPose on Jetson Nano and the inference is very slow and so I've been meaning to test the inference on MM Deploy since it supports Jetson's TensorRT.
> > > Could you please confirm if MM Pose is supported or not on MM Deploy?
> > 
> > 
> > Hi, I am trying to setup mmPose and mmdet in Jetson nano. I tried it in laptop. It worked fine. I think setting up both frameworks is totally different in Jetson nano as it uses aarch64. **Can you elaborate all steps for setting up mmdet and mmPose on Jetson nano. The help is much appreciated.**
> 
> You could refer to the doc: https://github.com/open-mmlab/mmdeploy/blob/master/docs/en/tutorials/how_to_install_mmdeploy_on_jetsons.md

Thanks fro the quick response. I just searched installing mmPose and mmDet on Jetson nano on google and came here. Can you give some idea about mmDeploy as I am not familiar with it? Can I use mmDeploy without installing mmPose and mmDet?

**Comment by RunningLeon** [comment]
> > > > I've been testing the MMDet and MMPose on Jetson Nano and the inference is very slow and so I've been meaning to test the inference on MM Deploy since it supports Jetson's TensorRT.
> > > > Could you please confirm if MM Pose is supported or not on MM Deploy?
> > > 
> > > 
> > > Hi, I am trying to setup mmPose and mmdet in Jetson nano. I tried it in laptop. It worked fine. I think setting up both frameworks is totally different in Jetson nano as it uses aarch64. **Can you elaborate all steps for setting up mmdet and mmPose on Jetson nano. The help is much appreciated.**
> > 
> > 
> > You could refer to the doc: https://github.com/open-mmlab/mmdeploy/blob/master/docs/en/tutorials/how_to_install_mmdeploy_on_jetsons.md
> 
> Thanks fro the quick response. I just searched installing mmPose and mmDet on Jetson nano on google and came here. Can you give some idea about mmDeploy as I am not familiar with it? Can I use mmDeploy without installing mmPose and mmDet?
> 
> what is the difference between direct mmPose & mmDet setup and using mmDeploy?

Hi, mmdeploy is a toolset to support the deployment of models from OpenMMLab codebases such mmcls, mmdet, mmpose. You have to install mmpose and mmdet through pip or build from source. Because mmdeploy has C code and you have to install it by building it yourself. Could refer to the doc https://github.com/open-mmlab/mmdeploy/blob/master/docs/en/tutorials/how_to_install_mmdeploy_on_jetsons.md

**Comment by Thevakumar-Luheerathan** [comment]
> > > > > I've been testing the MMDet and MMPose on Jetson Nano and the inference is very slow and so I've been meaning to test the inference on MM Deploy since it supports Jetson's TensorRT.
> > > > > Could you please confirm if MM Pose is supported or not on MM Deploy?
> > > > 
> > > > 
> > > > Hi, I am trying to setup mmPose and mmdet in Jetson nano. I tried it in laptop. It worked fine. I think setting up both frameworks is totally different in Jetson nano as it uses aarch64. **Can you elaborate all steps for setting up mmdet and mmPose on Jetson nano. The help is much appreciated.**
> > > 
> > > 
> > > You could refer to the doc: https://github.com/open-mmlab/mmdeploy/blob/master/docs/en/tutorials/how_to_install_mmdeploy_on_jetsons.md
> > 
> > 
> > Thanks fro the quick response. I just searched installing mmPose and mmDet on Jetson nano on google and came here. Can you give some idea about mmDeploy as I am not familiar with it? Can I use mmDeploy without installing mmPose and mmDet?
> > what is the difference between direct mmPose & mmDet setup and using mmDeploy?
> 
> Hi, mmdeploy is a toolset to support the deployment of models from OpenMMLab codebases such mmcls, mmdet, mmpose. You have to install mmpose and mmdet through pip or build from source. Because mmdeploy has C code and you have to install it by building it yourself. Could refer to the doc https://github.com/open-mmlab/mmdeploy/blob/master/docs/en/tutorials/how_to_install_mmdeploy_on_jetsons.md

Thank you
I will check on that

**Comment by Thevakumar-Luheerathan** [comment]
```
Complete output from command /usr/bin/python3 -c "import setuptools, tokenize;__file__='/home/nano4g/mmcv/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" develop --no-deps --user --prefix=:
running develop
running egg_info
writing mmcv_full.egg-info/PKG-INFO
writing dependency_links to mmcv_full.egg-info/dependency_links.txt
writing requirements to mmcv_full.egg-info/requires.txt
writing top-level names to mmcv_full.egg-info/top_level.txt
reading manifest file 'mmcv_full.egg-info/SOURCES.txt'
reading manifest template 'MANIFEST.in'
writing manifest file 'mmcv_full.egg-info/SOURCES.txt'
running build_ext
building 'mmcv._ext' extension
creating /home/nano4g/mmcv/build
creating /home/nano4g/mmcv/build/temp.linux-aarch64-3.6
creating /home/nano4g/mmcv/build/temp.linux-aarch64-3.6/mmcv
creating /home/nano4g/mmcv/build/temp.linux-aarch64-3.6/mmcv/ops
creating /home/nano4g/mmcv/build/temp.linux-aarch64-3.6/mmcv/ops/csrc
creating /home/nano4g/mmcv/build/temp.linux-aarch64-3.6/mmcv/ops/csrc/pytorch
creating /home/nano4g/mmcv/build/temp.linux-aarch64-3.6/mmcv/ops/csrc/pytorch/cpu
creating /home/nano4g/mmcv/build/temp.linux-aarch64-3.6/mmcv/ops/csrc/pytorch/cuda
Emitting ninja build file /home/nano4g/mmcv/build/temp.linux-aarch64-3.6/build.ninja...
Compiling objects...
Using envvar MAX_JOBS (4) as the number of workers...
[1/102] c++ -MMD -MF /home/nano4g/mmcv/build/temp.linux-aarch64-3.6/mmcv/ops/csrc/pytorch/pybind.o.d -pthread -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fPIC -DMMCV_WITH_CUDA -I/home/nano4g/mmcv/mmcv/ops/csrc/common -I/home/nano4g/mmcv/mmcv/ops/csrc/common/cuda -I/home/nano4g/.local/lib/python3.6/site-packages/torch/include -I/home/nano4g/.local/lib/python3.6/site-packages/torch/include/torch/csrc/api/include -I/home/nano4g/.local/lib/python3.6/site-packages/torch/include/TH -I/home/nano4g/.local/lib/python3.6/site-packages/torch/include/THC -I/usr/local/cuda/include -I/usr/include/python3.6m -c -c /home/nano4g/mmcv/mmcv/ops/csrc/pytorch/pybind.cpp -o /home/nano4g/mmcv/build/temp.linux-aarch64-3.6/mmcv/ops/csrc/pytorch/pybind.o -std=c++14 -DTORCH_API_INCLUDE_EXTENSION_H '-DPYBIND11_COMPILER_TYPE="_gcc"' '-DPYBIND11_STDLIB="_libstdcpp"' '-DPYBIND11_BUILD_ABI="_cxxabi1011"' -DTORCH_EXTENSION_NAME=_ext -D_GLIBCXX_USE_CXX11_ABI=1
FAILED: /home/nano4g/mmcv/build/temp.linux-aarch64-3.6/mmcv/ops/csrc/pytorch/pybind.o
c++ -MMD -MF /home/nano4g/mmcv/build/temp.linux-aarch64-3.6/mmcv/ops/csrc/pytorch/pybind.o.d -pthread -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fPIC -DMMCV_WITH_CUDA -I/home/nano4g/mmcv/mmcv/ops/csrc/common -I/home/nano4g/mmcv/mmcv/ops/csrc/common/cuda -I/home/nano4g/.local/lib/python3.6/site-packages/torch/include -I/home/nano4g/.local/lib/python3.6/site-packages/torch/include/torch/csrc/api/include -I/home/nano4g/.local/lib/python3.6/site-packages/torch/include/TH -I/home/nano4g/.local/lib/python3.6/site-packages/torch/include/THC -I/usr/local/cuda/include -I/usr/include/python3.6m -c -c /home/nano4g/mmcv/mmcv/ops/csrc/pytorch/pybind.cpp -o /home/nano4g/mmcv/build/temp.linux-aarch64-3.6/mmcv/ops/csrc/pytorch/pybind.o -std=c++14 -DTORCH_API_INCLUDE_EXTENSION_H '-DPYBIND11_COMPILER_TYPE="_gcc"' '-DPYBIND11_STDLIB="_libstdcpp"' '-DPYBIND11_BUILD_ABI="_cxxabi1011"' -DTORCH_EXTENSION_NAME=_ext -D_GLIBCXX_USE_CXX11_ABI=1
c++: internal compiler error: Killed (program cc1plus)
Please submit a full bug report,
with preprocessed source if appropriate.
See <file:///usr/share/doc/gcc-7/README.Bugs> for instructions.
[2/102] c++ -MMD -MF /home/nano4g/mmcv/build/temp.linux-aarch64-3.6/mmcv/ops/csrc/pytorch/border_align.o.d -pthread -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fPIC -DMMCV_WITH_CUDA -I/home/nano4g/mmcv/mmcv/ops/csrc/common -I/home/nano4g/mmcv/mmcv/ops/csrc/common/cuda -I/home/nano4g/.local/lib/python3.6/site-packages/torch/include -I/home/nano4g/.local/lib/python3.6/site-packages/torch/include/torch/csrc/api/include -I/home/nano4g/.local/lib/python3.6/site-packages/torch/include/TH -I/home/nano4g/.local/lib/python3.6/site-packages/torch/include/THC -I/usr/local/cuda/include -I/usr/include/python3.6m -c -c /home/nano4g/mmcv/mmcv/ops/csrc/pytorch/border_align.cpp -o /home/nano4g/mmcv/build/temp.linux-aarch64-3.6/mmcv/ops/csrc/pytorch/border_align.o -std=c++14 -DTORCH_API_INCLUDE_EXTENSION_H '-DPYBIND11_COMPILER_TYPE="_gcc"' '-DPYBIND11_STDLIB="_libstdcpp"' '-DPYBIND11_BUILD_ABI="_cxxabi1011"' -DTORCH_EXTENSION_NAME=_ext -D_GLIBCXX_USE_CXX11_ABI=1
FAILED: /home/nano4g/mmcv/build/temp.linux-aarch64-3.6/mmcv/ops/csrc/pytorch/border_align.o
c++ -MMD -MF /home/nano4g/mmcv/build/temp.linux-aarch64-3.6/mmcv/ops/csrc/pytorch/border_align.o.d -pthread -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fPIC -DMMCV_WITH_CUDA -I/home/nano4g/mmcv/mmcv/ops/csrc/common -I/home/nano4g/mmcv/mmcv/ops/csrc/common/cuda -I/home/nano4g/.local/lib/python3.6/site-packages/torch/include -I/home/nano4g/.local/lib/python3.6/site-packages/torch/include/torch/csrc/api/include -I/home/nano4g/.local/lib/python3.6/site-packages/torch/include/TH -I/home/nano4g/.local/lib/python3.6/site-packages/torch/include/THC -I/usr/local/cuda/include -I/usr/include/python3.6m -c -c /home/nano4g/mmcv/mmcv/ops/csrc/pytorch/border_align.cpp -o /home/nano4g/mmcv/build/temp.linux-aarch64-3.6/mmcv/ops/csrc/pytorch/border_align.o -std=c++14 -DTORCH_API_INCLUDE_EXTENSION_H '-DPYBIND11_COMPILER_TYPE="_gcc"' '-DPYBIND11_STDLIB="_libstdcpp"' '-DPYBIND11_BUILD_ABI="_cxxabi1011"' -DTORCH_EXTENSION_NAME=_ext -D_GLIBCXX_USE_CXX11_ABI=1
c++: internal compiler error: Killed (program cc1plus)
Please submit a full bug report,
with preprocessed source if appropriate.
See <file:///usr/share/doc/gcc-7/README.Bugs> for instructions.
[3/102] c++ -MMD -MF /home/nano4g/mmcv/build/temp.linux-aarch64-3.6/mmcv/ops/csrc/pytorch/fused_bias_leakyrelu.o.d -pthread -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fPIC -DMMCV_WITH_CUDA -I/home/nano4g/mmcv/mmcv/ops/csrc/common -I/home/nano4g/mmcv/mmcv/ops/csrc/common/cuda -I/home/nano4g/.local/lib/python3.6/site-packages/torch/include -I/home/nano4g/.local/lib/python3.6/site-packages/torch/include/torch/csrc/api/include -I/home/nano4g/.local/lib/python3.6/site-packages/torch/include/TH -I/home/nano4g/.local/lib/python3.6/site-packages/torch/include/THC -I/usr/local/cuda/include -I/usr/include/python3.6m -c -c /home/nano4g/mmcv/mmcv/ops/csrc/pytorch/fused_bias_leakyrelu.cpp -o /home/nano4g/mmcv/build/temp.linux-aarch64-3.6/mmcv/ops/csrc/pytorch/fused_bias_leakyrelu.o -std=c++14 -DTORCH_API_INCLUDE_EXTENSION_H '-DPYBIND11_COMPILER_TYPE="_gcc"' '-DPYBIND11_STDLIB="_libstdcpp"' '-DPYBIND11_BUILD_ABI="_cxxabi1011"' -DTORCH_EXTENSION_NAME=_ext -D_GLIBCXX_USE_CXX11_ABI=1
[4/102] c++ -MMD -MF /home/nano4g/mmcv/build/temp.linux-aarch64-3.6/mmcv/ops/csrc/pytorch/nms_rotated.o.d -pthread -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fPIC -DMMCV_WITH_CUDA -I/home/nano4g/mmcv/mmcv/ops/csrc/common -I/home/nano4g/mmcv/mmcv/ops/csrc/common/cuda -I/home/nano4g/.local/lib/python3.6/site-packages/torch/include -I/home/nano4g/.local/lib/python3.6/site-packages/torch/include/torch/csrc/api/include -I/home/nano4g/.local/lib/python3.6/site-packages/torch/include/TH -I/home/nano4g/.local/lib/python3.6/site-packages/torch/include/THC -I/usr/local/cuda/include -I/usr/include/python3.6m -c -c /home/nano4g/mmcv/mmcv/ops/csrc/pytorch/nms_rotated.cpp -o /home/nano4g/mmcv/build/temp.linux-aarch64-3.6/mmcv/ops/csrc/pytorch/nms_rotated.o -std=c++14 -DTORCH_API_INCLUDE_EXTENSION_H '-DPYBIND11_COMPILER_TYPE="_gcc"' '-DPYBIND11_STDLIB="_libstdcpp"' '-DPYBIND11_BUILD_ABI="_cxxabi1011"' -DTORCH_EXTENSION_NAME=_ext -D_GLIBCXX_USE_CXX11_ABI=1
ninja: build stopped: subcommand failed.
Traceback (most recent call last):
  File "/home/nano4g/.local/lib/python3.6/site-packages/torch/utils/cpp_extension.py", line 1723, in _run_ninja_build
    env=env)
  File "/usr/lib/python3.6/subprocess.py", line 438, in run
    output=stdout, stderr=stderr)
subprocess.CalledProcessError: Command '['ninja', '-v', '-j', '4']' returned non-zero exit status 1.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/nano4g/mmcv/setup.py", line 390, in <module>
    zip_safe=False)
  File "/usr/lib/python3/dist-packages/setuptools/__init__.py", line 129, in setup
    return distutils.core.setup(**attrs)
  File "/usr/lib/python3.6/distutils/core.py", line 148, in setup
    dist.run_commands()
  File "/usr/lib/python3.6/distutils/dist.py", line 955, in run_commands
    self.run_command(cmd)
  File "/usr/lib/python3.6/distutils/dist.py", line 974, in run_command
    cmd_obj.run()
  File "/usr/lib/python3/dist-packages/setuptools/command/develop.py", line 36, in run
    self.install_for_development()
  File "/usr/lib/python3/dist-packages/setuptools/command/develop.py", line 136, in install_for_development
    self.run_command('build_ext')
  File "/usr/lib/python3.6/distutils/cmd.py", line 313, in run_command
    self.distribution.run_command(command)
  File "/usr/lib/python3.6/distutils/dist.py", line 974, in run_command
    cmd_obj.run()
  File "/usr/lib/python3/dist-packages/setuptools/command/build_ext.py", line 78, in run
    _build_ext.run(self)
  File "/home/nano4g/.local/lib/python3.6/site-packages/Cython/Distutils/old_build_ext.py", line 186, in run
    _build_ext.build_ext.run(self)
  File "/usr/lib/python3.6/distutils/command/build_ext.py", line 339, in run
    self.build_extensions()
  File "/home/nano4g/.local/lib/python3.6/site-packages/torch/utils/cpp_extension.py", line 735, in build_extensions
    build_ext.build_extensions(self)
  File "/home/nano4g/.local/lib/python3.6/site-packages/Cython/Distutils/old_build_ext.py", line 195, in build_extensions
    _build_ext.build_ext.build_extensions(self)
  File "/usr/lib/python3.6/distutils/command/build_ext.py", line 448, in build_extensions
    self._build_extensions_serial()
  File "/usr/lib/python3.6/distutils/command/build_ext.py", line 473, in _build_extensions_serial
    self.build_extension(ext)
  File "/usr/lib/python3/dist-packages/setuptools/command/build_ext.py", line 199, in build_extension
    _build_ext.build_extension(self, ext)
  File "/usr/lib/python3.6/distutils/command/build_ext.py", line 533, in build_extension
    depends=ext.depends)
  File "/home/nano4g/.local/lib/python3.6/site-packages/torch/utils/cpp_extension.py", line 565, in unix_wrap_ninja_compile
    with_cuda=with_cuda)
  File "/home/nano4g/.local/lib/python3.6/site-packages/torch/utils/cpp_extension.py", line 1404, in _write_ninja_file_and_compile_objects
    error_prefix='Error compiling objects for extension')
  File "/home/nano4g/.local/lib/python3.6/site-packages/torch/utils/cpp_extension.py", line 1733, in _run_ninja_build
    raise RuntimeError(message) from e
RuntimeError: Error compiling objects for extension

----------------------------------------
```

Hi @RunningLeon I got a issue as this in Jetson nano while compiling using ninja. How to solve this issue?

**Comment by RunningLeon** [comment]
@Thevakumar-Luheerathan Hi, you may ask in [mmcv](https://github.com/open-mmlab/mmcv) repo for more help. As far as I know, there is no such wheel by far due to the limited use and hard maintenance.

**Comment by Thevakumar-Luheerathan** [comment]
> 

ok Thank you.

**Comment by RunningLeon** [comment]
closed since #94 is merged.
